src/app/convai-checker.component.spec.ts

```diff
@@ -1455,4 +1500,394 @@ describe('Convai checker test', () => {
+
+  it('Test circle square diamond change for score thresholds', async(() => {
+    let fixture = TestBed.createComponent(ConvaiCheckerCustomDemoSettingsTestComponent);
+
+    // Configure settings.
+    let demoSettings = getCopyOfDefaultDemoSettings();
+    demoSettings.scoreThresholds = [0, 0.6, 0.8];
+    demoSettings.loadingIconStyle = LoadingIconStyle.CIRCLE_SQUARE_DIAMOND;
+    fixture.componentInstance.setDemoSettings(demoSettings);
+
+    fixture.detectChanges();
+    let checker = fixture.componentInstance.checker;
+
+    // Set up the mock responses for the series of three requests that will be
+    // made in the test.
+    let queryTexts = [
+      'Your mother was a hamster',
+      'Your father smelled of elderberries',
+      'What is the air velocity of an unladen swallow?'
+    ];
+
+    let callCount = 0;
+    let mockResponses = [
+      getMockCheckerResponseWithScore(0.9, checker.getToken(queryTexts[0])),
+      getMockCheckerResponseWithScore(0.7, checker.getToken(queryTexts[1])),
+      getMockCheckerResponseWithScore(0.2, checker.getToken(queryTexts[2]))
+    ];
+
+    let mockBackend = TestBed.get(MockBackend);
+    mockBackend.connections
+     .subscribe((connection: MockConnection) => {
+       fixture.detectChanges();
+       // Check the UI state before returning the repsonse.
+       verifyCircleSquareDiamondWidgetVisible();
+       expect(checker.statusWidget.isLoading).toBe(true);
+       connection.mockRespond(
+         new Response(
+           new ResponseOptions({
+              body: mockResponses[callCount]
+           })
+         )
+       );
+
+       // Wait for async code to complete.
+       fixture.whenStable().then(() => {
+         fixture.detectChanges();
+         // Checks the UI state after the response has been received.
+
+         // Checks that loading has stopped.
+         expect(checker.statusWidget.isLoading).toBe(false);
+         expect(checker.statusWidget.isPlayingLoadingAnimation).toBe(false);
+
+         // For each of the mock responses, the thresholds should indicate a
+         // different loading icon shape.
+         let expectedShape: Shape;
+         if (callCount === 0) {
+           expectedShape = Shape.DIAMOND;
+         } else if (callCount === 1) {
+           expectedShape = Shape.SQUARE;
+         } else {
+           expectedShape = Shape.CIRCLE;
+         }
+         verifyLoadingWidgetHasShape(checker, expectedShape);
+
+         if (callCount < 2) {
+           callCount++;
+           // Fire another request.
+           setTextAndFireInputEvent(queryTexts[callCount], textArea);
+         }
+       });
+    });
+
+    verifyCircleSquareDiamondWidgetVisible();
+
+    let textArea = fixture.debugElement.query(
+      By.css('#' + checker.inputId)).nativeElement;
+
+    // Send an input event to trigger the service call.
+    setTextAndFireInputEvent(queryTexts[callCount], textArea);
+  }));
+
+  it('Test emoji change for score thresholds', async(() => {
```

Comment:
Literally the exact same comments as above for this.

Also, again on the note of parameterized tests, these could be combined into the above with just the extra inclusion of a parameter of `loadingIconStyle`.

Reply:
Noted; for now I'm going to make the decision keep these separated, though, since there could potentially be independent changes made in each one later. Also, parameterized tests might result in a cost of reduced readability in exchange for less code in this case, because I'd end up with some ugly ternary expressions in the tests (for which visibility function to check, etc) and I'd rather not have that.